File: server/src/test/java/com/rocketmq/studio/instance/acl/AclServiceTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.rocketmq.studio.instance.acl;

import com.rocketmq.studio.common.exception.BusinessException;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
class AclServiceTest {

    @Mock
    private AclRepository aclRepository;

    @InjectMocks
    private AclService aclService;

    @Test
    void listRulesShouldReturnRulesFromRepository() {
        List<AclRuleVO> rules = List.of(
                AclRuleVO.builder().principal("user1").resource("topic-1").decision("ALLOW").build(),
                AclRuleVO.builder().principal("user2").resource("topic-2").decision("DENY").build()
        );
        when(aclRepository.findRules("cluster-1", "user1")).thenReturn(rules);

        List<AclRuleVO> result = aclService.listRules("cluster-1", "user1");

        assertThat(result).hasSize(2);
        assertThat(result.get(0).getPrincipal()).isEqualTo("user1");
        verify(aclRepository).findRules("cluster-1", "user1");
    }

    @Test
    void listRulesShouldPassNullFilters() {
        when(aclRepository.findRules(null, null)).thenReturn(List.of());

        List<AclRuleVO> result = aclService.listRules(null, null);

        assertThat(result).isEmpty();
        verify(aclRepository).findRules(null, null);
    }

    @Test
    void createRuleShouldSetIdAndTimestamp() {
        AclRuleVO input = AclRuleVO.builder()
                .principal("user1")
                .resource("topic-1")
                .resourceType("TOPIC")
                .decision("ALLOW")
                .build();

        when(aclRepository.saveRule(any(AclRuleVO.class))).thenAnswer(inv -> inv.getArgument(0));

        AclRuleVO result = aclService.createRule(input);

        assertThat(result.getId()).isNotBlank();
        assertThat(result.getCreatedAt()).isNotNull();
        assertThat(result.getPrincipal()).isEqualTo("user1");
        assertThat(result.getResource()).isEqualTo("topic-1");
        verify(aclRepository).saveRule(any(AclRuleVO.class));
    }

    @Test
    void createRuleShouldRequirePrincipal() {
        AclRuleVO input = AclRuleVO.builder()
                .principal(" ")
                .resource("topic-1")
                .build();

        assertThatThrownBy(() -> aclService.createRule(input))
                .isInstanceOf(BusinessException.class)
                .satisfies(ex -> assertThat(((BusinessException) ex).getCode()).isEqualTo(400))
                .hasMessage("ACL principal is required");
        verify(aclRepository, never()).saveRule(any(AclRuleVO.class));
    }

    @Test
    void createRuleShouldRequireResource() {
        AclRuleVO input = AclRuleVO.builder()
                .principal("user1")
                .resource(" ")
                .build();

        assertThatThrownBy(() -> aclService.createRule(input))
                .isInstanceOf(BusinessException.class)
                .satisfies(ex -> assertThat(((BusinessException) ex).getCode()).isEqualTo(400))
                .hasMessage("ACL resource is required");
        verify(aclRepository, never()).saveRule(any(AclRuleVO.class));
    }

    @Test
    void deleteRuleShouldDelegateToRepository() {
        aclService.deleteRule("rule-1");

        verify(aclRepository).deleteRule("rule-1");
    }

    @Test
    void updateRuleShouldRequireId() {
        AclRuleVO input = AclRuleVO.builder()
                .principal("user1")
                .resource("topic-1")
                .build();

        assertThatThrownBy(() -> aclService.updateRule(input))
                .hasMessage("ACL rule id is required");
    }

    @Test
    void updateRuleShouldSaveExistingRule() {
        AclRuleVO input = AclRuleVO.builder()
                .id("rule-1")
                .principal("user1")
                .resource("topic-1")
                .decision("DENY")
                .build();

        when(aclRepository.saveRule(any(AclRuleVO.class))).thenAnswer(inv -> inv.getArgument(0));

        AclRuleVO result = aclService.updateRule(input);

        assertThat(result.getId()).isEqualTo("rule-1");
        assertThat(result.getCreatedAt()).isNotNull();
        assertThat(result.getDecision()).isEqualTo("DENY");
        verify(aclRepository).saveRule(any(AclRuleVO.class));
    }

    @Test
    void listUsersShouldReturnAllUsers() {
        List<AclUserVO> users = List.of(
                AclUserVO.builder().username("admin").admin(true).build(),
                AclUserVO.builder().username("reader").admin(false).build()
        );
        when(aclRepository.findUsers()).thenReturn(users);

        List<AclUserVO> result = aclService.listUsers();

        assertThat(result).hasSize(2);
        assertThat(result.get(0).getUsername()).isEqualTo("admin");
        verify(aclRepository).findUsers();
    }

    @Test
    void createUserShouldGenerateKeysAndSetTimestamp() {
        AclUserVO input = AclUserVO.builder()
                .username("newuser")
                .admin(false)
                .build();

        when(aclRepository.saveUser(any(AclUserVO.class))).thenAnswer(inv -> inv.getArgument(0));

        AclUserVO result = aclService.createUser(input);

        assertThat(result.getId()).isNotBlank();
        assertThat(result.getAccessKey()).isNotBlank();
        assertThat(result.getSecretKey()).isNotBlank();
        assertThat(result.getAccessKey()).doesNotContain("-");
        assertThat(result.getSecretKey()).doesNotContain("-");
        assertThat(result.getCreatedAt()).isNotNull();
        assertThat(result.getUsername()).isEqualTo("newuser");
        verify(aclRepository).saveUser(any(AclUserVO.class));
    }

    @Test
    void createUserShouldRequireUsername() {
        AclUserVO input = AclUserVO.builder()
                .username(" ")
                .admin(false)
                .build();

        assertThatThrownBy(() -> aclService.createUser(input))
                .isInstanceOf(BusinessException.class)
                .satisfies(ex -> assertThat(((BusinessException) ex).getCode()).isEqualTo(400))
                .hasMessage("ACL username is required");
        verify(aclRepository, never()).saveUser(any(AclUserVO.class));
    }

    @Test
    void deleteUserShouldDelegateToRepository() {
        aclService.deleteUser("user-1");

        verify(aclRepository).deleteUser("user-1");
    }

    @Test
    void updateUserShouldRequireId() {
        AclUserVO input = AclUserVO.builder()
                .username("newuser")
                .build();

        assertThatThrownBy(() -> aclService.updateUser(input))
                .hasMessage("ACL user id is required");
    }

    @Test
    void updateUserShouldSaveExistingUser() {
        AclUserVO input = AclUserVO.builder()
                .id("user-1")
                .username("newuser")
                .accessKey("ak")
                .secretKey("sk")
                .admin(true)
                .build();

        when(aclRepository.saveUser(any(AclUserVO.class))).thenAnswer(inv -> inv.getArgument(0));

        AclUserVO result = aclService.updateUser(input);

        assertThat(result.getId()).isEqualTo("user-1");
        assertThat(result.getCreatedAt()).isNotNull();
        assertThat(result.isAdmin()).isTrue();
        verify(aclRepository).saveUser(any(AclUserVO.class));
    }
}

File: web/src/services/aclService.test.ts
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { describe, expect, it, vi } from 'vitest';
import {
  createAclRule,
  createAclUser,
  createAndUpdatePlainAccessConfig,
  deleteAclRule,
  deleteAclUser,
  examineBrokerClusterAclConfig,
  getAclUserCredentials,
  listAclRules,
  listAclUsers,
  pageAclUsers,
  updateAclRule,
  updateAclUser,
} from './aclService';

vi.mock('./dataMode', () => ({ isMockMode: () => true }));
vi.mock('../config', () => ({
  API_BASE_URL: '/api',
}));

describe('ACL service mock data', () => {
  it('returns copied ACL rule rows', async () => {
    const first = await listAclRules({ principal: 'user-admin' });
    expect(first.items[0].principal).toBe('user-admin');

    first.items[0].principal = 'mutated-principal';
    first.items[0].actions.push('MUTATED');

    const second = await listAclRules({ principal: 'user-admin' });
    expect(second.items[0].principal).toBe('user-admin');
    expect(second.items[0].actions).toEqual(['ALL']);
    expect(second.items[0]).not.toBe(first.items[0]);
  });

  it('copies ACL rule arrays on create and update', async () => {
    const actions = ['PUB'];
    const created = await createAclRule({
      principal: 'user-created-copy-test',
      resource: 'created-topic',
      actions,
    });
    actions.push('SUB');
    created.actions.push('MUTATED');

    const afterCreate = await listAclRules({ principal: 'user-created-copy-test' });
    expect(afterCreate.items[0].actions).toEqual(['PUB']);

    const updateActions = ['SUB'];
    const updated = await updateAclRule({ id: created.id, actions: updateActions });
    updateActions.push('PUB');
    updated.actions.push('MUTATED');

    const afterUpdate = await listAclRules({ principal: 'user-created-copy-test' });
    expect(afterUpdate.items[0].actions).toEqual(['SUB']);
  });

  it('returns copied ACL user rows', async () => {
    const first = await listAclUsers({ keyword: 'user-admin' });
    expect(first[0].username).toBe('user-admin');

    first[0].username = 'mutated-user';
    first[0].clusters.push('mutated-cluster');

    const second = await listAclUsers({ keyword: 'user-admin' });
    expect(second[0].username).toBe('user-admin');
    expect(second[0].clusters).not.toContain('mutated-cluster');
    expect(second[0]).not.toBe(first[0]);
  });

  it('validates ACL user pagination like the backend', async () => {
    await expect(pageAclUsers({ page: 0, pageSize: 20 })).rejects.toThrow('page must be >= 1');
    await expect(pageAclUsers({ page: 1, pageSize: 101 })).rejects.toThrow(
      'pageSize must be between 1 and 100',
    );
  });

  it('copies ACL user arrays on create and update', async () => {
    const clusters = ['rmq-created'];
    const created = await createAclUser({
      username: 'user-created-copy-test',
      clusters,
    });
    clusters.push('rmq-mutated');
    created.clusters.push('rmq-mutated-return');

    const afterCreate = await listAclUsers({ keyword: 'user-created-copy-test' });
    expect(afterCreate[0].clusters).toEqual(['rmq-created']);

    const updateClusters = ['rmq-updated'];
    const updated = await updateAclUser({ id: created.id, clusters: updateClusters });
    updateClusters.push('rmq-mutated');
    updated.clusters.push('rmq-mutated-return');

    const afterUpdate = await listAclUsers({ keyword: 'user-created-copy-test' });
    expect(afterUpdate[0].clusters).toEqual(['rmq-updated']);
  });

  it('builds cluster ACL config from mock accounts', async () => {
    const config = await examineBrokerClusterAclConfig('DefaultCluster');
    expect(config.aclEnabled).toBe(true);
    expect(config.aclVersion).toBe('ACL 2.0');
    expect(config.globalWhiteRemoteAddresses).toContain('192.168.0.0/16');
    expect(config.accountCount).toBe(config.accounts.length);
    expect(config.accounts[0].accessKey).toBe('user-admin');
  });

  it('creates and updates a plain access account in mock state', async () => {
    const created = await createAndUpdatePlainAccessConfig({
      accessKey: 'svc-mock',
      secretKey: 'svc-mock-secret-value',
      admin: false,
      defaultTopicPerm: 'PUB',
      topicPerms: ['a=PUB'],
    });
    expect(created.accessKey).toBe('svc-mock');
    // The secret is echoed only when it was just provided.
    expect(created.secretKey).toBe('svc-mock-secret-value');

    const updated = await createAndUpdatePlainAccessConfig({
      accessKey: 'svc-mock',
      admin: true,
      defaultTopicPerm: 'ALL',
    });
    expect(updated.admin).toBe(true);
    // A blank secret keeps the stored one and is not echoed back.
    expect(updated.secretKey).toBeNull();

    const config = await examineBrokerClusterAclConfig('c');
    const account = config.accounts.find((a) => a.accessKey === 'svc-mock');
    expect(account && account.admin).toBe(true);
    // Read-back views mask the secret instead of exposing the plaintext.
    expect(account?.secretKey).not.toBe('svc-mock-secret-value');
    expect(account?.secretKey).toContain('****');
  });

  it('rejects a new plain access account without a secret', async () => {
    await expect(createAndUpdatePlainAccessConfig({ accessKey: 'svc-no-secret' })).rejects.toThrow(
      'secretKey is required',
    );
  });

  it('removes a rule from mock state and tolerates an unknown id', async () => {
    const created = await createAclRule({
      principal: 'user-delete-candidate',
      resource: 'delete-me',
      actions: ['PUB'],
    });
    expect(created.id).toBeGreaterThan(0);

    await deleteAclRule(created.id);

    const after = await listAclRules({ principal: 'user-delete-candidate' });
    expect(after.total).toBe(0);
    await expect(deleteAclRule(99999999)).resolves.toBeUndefined();
  });

  it('removes a user from mock state and tolerates an unknown id', async () => {
    const created = await createAclUser({
      username: 'user-delete-candidate',
      clusters: ['rmq-delete-test'],
    });

    await deleteAclUser(created.id);

    const after = await listAclUsers({ keyword: 'user-delete-candidate' });
    expect(after).toHaveLength(0);
    await expect(deleteAclUser(99999999)).resolves.toBeUndefined();
  });

  it('returns a defensive copy of stored credentials for a known user', async () => {
    const credentials = await getAclUserCredentials(1);
    expect(credentials.username).toBe('user-admin');
    expect(credentials.secretKey).toBe('HqWz****xK8P');

    credentials.clusters.push('mutated-cluster');
    const again = await getAclUserCredentials(1);
    expect(again.clusters).not.toContain('mutated-cluster');
    expect(again).not.toBe(credentials);
  });

  it('throws for credentials of an unknown user', async () => {
    await expect(getAclUserCredentials(99999999)).rejects.toThrow('ACL user not found');
  });

  it('throws when updating a rule or user that does not exist', async () => {
    await expect(
      updateAclRule({ id: 99999999, principal: 'user-missing' }),
    ).rejects.toThrow('ACL rule not found');
    await expect(
      updateAclUser({ id: 99999999, username: 'user-missing' }),
    ).rejects.toThrow('ACL user not found');
  });

  it('filters rules by scope, decision, ACL version, and resource', async () => {
    const namespaced = await listAclRules({ scope: 'namespace' });
    expect(namespaced.items.length).toBeGreaterThan(0);
    expect(namespaced.items.every((rule) => rule.scope === 'namespace')).toBe(true);

    const denied = await listAclRules({ decision: 'DENY' });
    expect(denied.items.length).toBeGreaterThan(0);
    expect(denied.items.every((rule) => rule.decision === 'DENY')).toBe(true);

    const v1 = await listAclRules({ aclVersion: '1.0' });
    expect(v1.items.length).toBeGreaterThan(0);
    expect(v1.items.every((rule) => String(rule.aclVersion) === '1.0')).toBe(true);

    const resource = await listAclRules({ resource: 'payment' });
    expect(resource.items.length).toBeGreaterThan(0);
    expect(resource.items.every((rule) => rule.resource.toLowerCase().includes('payment'))).toBe(
      true,
    );
  });

  it('filters rules case-insensitively by principal', async () => {
    const lower = await listAclRules({ principal: 'user-admin' });
    const upper = await listAclRules({ principal: 'USER-ADMIN' });
    expect(lower.items.length).toBeGreaterThan(0);
    expect(upper.items).toEqual(lower.items);
  });

  it('pages rule results with a page size and total', async () => {
    const first = await listAclRules({ page: 1, pageSize: 3 });
    expect(first.items).toHaveLength(3);
    expect(first.total).toBeGreaterThan(3);
    expect(first.page).toBe(1);

    const second = await listAclRules({ page: 2, pageSize: 3 });
    expect(second.items).toHaveLength(3);
    expect(second.items).not.toEqual(first.items);
    expect(second.total).toBe(first.total);
  });

  it('filters users by a case-insensitive keyword', async () => {
    const upper = await listAclUsers({ keyword: 'USER-ADMIN' });
    expect(upper[0].username).toBe('user-admin');

    const partial = await listAclUsers({ keyword: 'order' });
    expect(partial.every((user) => user.username.toLowerCase().includes('order'))).toBe(true);
    expect(partial.some((user) => user.username === 'user-order-service')).toBe(true);
  });
});
